src/sqlfluff/rules/L051.py

```diff
@@ -0,0 +1,60 @@
+"""Implementation of Rule L051."""
+from typing import Optional
+from sqlfluff.core.parser.segments.raw import KeywordSegment, WhitespaceSegment
+
+from sqlfluff.core.rules.base import BaseRule, LintResult, LintFix, RuleContext
+from sqlfluff.core.rules.doc_decorators import document_fix_compatible
+
+
+@document_fix_compatible
+class Rule_L051(BaseRule):
+    """INNER JOIN must be fully qualified.
+
+    | **Anti-pattern**
+    | Lone JOIN is used.
+
+    .. code-block:: sql
+       :force:
+
+        SELECT
+            foo
+        FROM bar
+        JOIN baz;
+
+    | **Best practice**
+    | Use INNER JOIN rather than JOIN.
+
+    .. code-block:: sql
+       :force:
+
+        SELECT
+            foo
+        FROM bar
+        INNER JOIN baz;
+    """
+
+    def _eval(self, context: RuleContext) -> Optional[LintResult]:
+        """INNER JOIN must be fully qualified."""
+        # We are only interested in JOIN clauses.
+        if context.segment.type != "join_clause":
+            return None
+
+        # We identify non-lone JOIN by looking at first child segment.
+        if context.segment.segments[0].name != "join":
+            return None
+
+        # Replace lone JOIN with INNER JOIN.
+        return LintResult(
+            context.segment.segments[0],
+            fixes=[
+                LintFix(
+                    "edit",
```

Comment:
I think we want to use `"create"` here rather than `"edit"`.

Reason: `"create"` adds new **sibling** segments, while `"edit"` creates a new level of the parse tree, replacing the specified segment with a new segment of the same type, with the new segments _as children_. This is a subtle bug, because the final SQL **string** looks okay, but the parse **tree** is wrong. This can cause other lint rules to misbehave.

See an excerpt of `sqlfluff parse` output below for two queries, one using `JOIN`, the other using `INNER JOIN`. In the second, you'll see that `INNER` and the whitespace should be at the same level as the `JOIN`.

```
[L:  1, P: 30]      |                    [META] dedent:
[L:  1, P: 30]      |                    join_clause:
[L:  1, P: 30]      |                        keyword:                              'JOIN'
[L:  1, P: 34]      |                        [META] indent:
[L:  1, P: 34]      |                        whitespace:                           ' '
[L:  1, P: 35]      |                        from_expression_element:
[L:  1, P: 35]      |                            table_expression:
[L:  1, P: 35]      |                                table_reference:
[L:  1, P: 35]      |                                    identifier:               'bar'
[L:  1, P: 38]      |                        [META] dedent:
[L:  1, P: 38]      |    newline:                                                  '\n'
```

```
[L:  1, P: 30]      |                    [META] dedent:
[L:  1, P: 30]      |                    join_clause:
[L:  1, P: 30]      |                        keyword:                              'INNER'
[L:  1, P: 35]      |                        whitespace:                           ' '
[L:  1, P: 36]      |                        keyword:                              'JOIN'
[L:  1, P: 40]      |                        [META] indent:
[L:  1, P: 40]      |                        whitespace:                           ' '
[L:  1, P: 41]      |                        from_expression_element:
[L:  1, P: 41]      |                            table_expression:
[L:  1, P: 41]      |                                table_reference:
[L:  1, P: 41]      |                                    identifier:               'bar'
[L:  1, P: 44]      |                        [META] dedent:
[L:  1, P: 44]      |    newline:                                                  '\n'
```

Reply:
This rule is similar to L033. I see it has the same issue. I'll look into creating a PR to fix that one.

Reply:
@barrywhart Interesting, I wasn't aware of that! Will bear it in mind going forwards.

I've made the switch to using a create LintFix now, unit tests are all passing still but as you said this should prevent future bugs 😄

Reply:
Thanks! The behavior I described is pretty tricky and not documented. I hope in the future, we can add automated checking inside the linter to flag it if a rule does something to the parse tree that makes it illegal wrt the dialect.